assignments/assignment02.rb

```diff
@@ -73,3 +111,167 @@ def pluck(ary, key)
+<hr>
+<table>
+<tr><td>Starting Balance:</td><td align=right>$#{starting_balance}</td></tr>
+<tr><td>Total Deposits:</td><td align=right>#{total_deposits}</td></tr>
+<tr><td>Total Withdrawals:</td><td align=right>#{total_withdrawals}</td></tr>
+<tr><td>Ending Balance:</td><td align=right>$#{end_balance}</td></tr>
+</table>
+  EOM
+end
+
+def html_transaction(date, source, amount)
+  amount < 0 ? color = "red" : color = "black"
+  <<-EOM
+<tr><td>#{date}</td><td><i>#{source}</i></td><td align=right><font color=#{color}>$#{amount}</font></td><td></td></tr>
+  EOM
+end
+
+def html_table_header(title)
+  <<-EOM
+<h2>#{title}</h2>
+<hr>
+<table>
+<tr>
+<td align=left valign=bottom><b>Date</b></td>
+<td align=left valign=bottom><b>Source</b></td>
+<td align=right valign=bottom><b>Amount</b>
+</tr>
+  EOM
+end
+
+def html_daily_header
+  <<-EOM
+<h2>Daily Transactions with Balances:</h2>
+<hr>
+<table>
+<tr>
+<td align=left valign=bottom><b>Date</b></td>
+<td align=left valign=bottom><b>Source</b></td>
+<td align=right valign=bottom><b>Amount</b>
+</td><td align=right valign=bottom><b>Daily<br>Balance</b></td>
+</tr>
+  EOM
+end
+
+def html_table_footer
+  <<-EOM
+</table>
+  EOM
+end
+
+def html_daily_total(amount)
+  amount < 0 ? color = "red" : color = "black"
+  <<-EOM
+<tr><td colspan=3></td><td align=right><b><font color=#{color}>$#{amount}</font></b></td></tr>
+  EOM
+end
+
+
+starting_balance = 0
+
+fh = File.open("assignment02-input.csv")
+csv_data = fh.read()
+fh.close
+
+column_headers = Array.new
+transactions = Hash.new
+
+csv_data.each_line.map do |line|
+    entry = line.split(",")
+    if entry[0].include?("date")
+      # These are headers
+      column_headers = entry
+    else
+      # This is data
+      month, day, year = entry[0].split("/")
+      month = month.to_s.rjust(2, "0")
+      day = day.to_s.rjust(2, "0")
+      key = "#{year}-#{month}-#{day}"
+      source = entry[1]
+      entry[3].include?("withdrawal") ? amount = entry[2].to_f * -1 : amount = entry[2].to_f
+      transactions[key] = [] if transactions[key].nil?
+      transactions[key] << [ source, amount ]
+    end
+end
+
+transactions = transactions.sort
+
+puts html_header
+
+running_balance = starting_balance
+
+total_deposits = 0
+total_withdrawals = 0
+
+# - withdrawals
+puts html_table_header("List of Withdrawals:")
+transactions.each do |key, value|
+  year, month, day = key.split("-")
+  value.each do |transaction|
+    source = transaction[0]
+    amount = transaction[1]
+    puts html_transaction("#{month}/#{day}/#{year}", source, amount) if  amount < 0
+  end
+end
+puts html_table_footer
+
+# - deposits
+puts html_table_header("List of Deposits:")
+transactions.each do |key, value|
+  year, month, day = key.split("-")
+  value.each do |transaction|
+    source = transaction[0]
+    amount = transaction[1]
+    puts html_transaction("#{month}/#{day}/#{year}", source, amount) if amount >= 0
+  end
+end
+puts html_table_footer
+
+# - daily balance
```

Comment:
You are listing each transaction with a running total, but you aren't grouping transactions by date and calculating the final balance for the date.